### ISSUE TITLE
SERVFAIL to queries from localhost

### ISSUE THREAD

**Issue by saloniamatteo** [NLnetLabs/unbound#1172]
**Describe the bug**
Greetings,
Unbound answers `SERVFAIL` to **ALL** queries from localhost (`127.0.0.1`, `::1`, `192.168.1.88`).

No matter which settings I modify (including `interfaces:`), only queries to my external IP actually contain an answer...

(For context, what I really need is for port 8080 (HTTPS) to be available ONLY for localhost, however, that seems to not be the case...)

**To reproduce**
Relevant `/etc/unbound/unbound.conf`:

```conf
server:
	interface: lo@8080
	interface: eth0@53
	interface: eth0@853
	port: 53

	prefer-ip4: yes
	outgoing-port-avoid: "3200-3208"
	so-reuseport: yes
	udp-connect: yes
	cache-min-ttl: 60

	access-control: 127.0.0.1 allow
	access-control: ::1 allow
	access-control: ::ffff:127.0.0.1 allow
	access-control: 192.168.1.0/24 allow

	root-hints: "/etc/unbound/root.hints"
	hide-trustanchor: yes
	harden-dnssec-stripped: yes
	qname-minimisation: yes

	do-not-query-localhost: no

	prefetch: yes
	prefetch-key: yes

	auto-trust-anchor-file: "/etc/unbound/root-anchors.txt"
	trust-anchor-signaling: yes
	root-key-sentinel: yes
```

- Run `dig @localhost example.com` (or any other internal IP address)

- See the following:

```bind
;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;example.com.                   IN      A

;; Query time: 3 msec
;; SERVER: ::1#53(::1) (UDP)
;; WHEN: Sat Nov 02 00:49:25 CET 2024
;; MSG SIZE  rcvd: 40
```

**Expected behavior**
Running `dig @localhost example.com` should return valid results, just like how `dig @(external IP) example.com` does already:

```bind
;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1220
; COOKIE: e383007f25ecdecf3acbbe4967256b627c73bec4a688ea26 (good)
;; QUESTION SECTION:
;example.com.                   IN      A

;; ANSWER SECTION:
example.com.            1514    IN      A       93.184.215.14

;; Query time: 31 msec
;; SERVER: (external IP)#53((external IP)) (UDP)
;; WHEN: Sat Nov 02 00:59:30 CET 2024
;; MSG SIZE  rcvd: 84
```

**System:**
 - Unbound version: 1.20.0
 - OS: Gentoo Linux 6.11.3-gentoo
 - `unbound -V` output:

```
Version 1.20.0

Configure line: --prefix=/usr --build=x86_64-pc-linux-gnu --host=x86_64-pc-linux-gnu --mandir=/usr/share/man --infodir=/usr/share/info --datadir=/usr/share --sysconfdir=/etc --localstatedir=/var/lib --datarootdir=/usr/share --disable-static --docdir=/usr/share/doc/unbound-1.20.0 --htmldir=/usr/share/doc/unbound-1.20.0/html --with-sysroot=/ --libdir=/usr/lib64 --disable-debug --disable-gost --disable-dnscrypt --disable-dnstap --enable-ecdsa --disable-subnet --disable-cachedb --disable-static --disable-systemd --without-pythonmodule --without-pyunbound --with-pthreads --with-libnghttp2 --enable-tfo-client --enable-tfo-server --disable-flto --disable-rpath --enable-event-api --enable-ipsecmod --with-libevent=/usr --without-libhiredis --with-pidfile=/run/unbound.pid --with-rootkey-file=/etc/dnssec/root-anchors.txt --with-ssl=/usr --with-libexpat=/usr
Linked libs: libevent 2.1.12-stable (it uses epoll), OpenSSL 3.2.2 4 Jun 2024
Linked modules: dns64 ipsecmod respip validator iterator
TCP Fastopen feature available

BSD licensed, see LICENSE in source package for details.
Report bugs to unbound-bugs@nlnetlabs.nl or https://github.com/NLnetLabs/unbound/issues
```

**Additional information**
(`dig @localhost example.com`, SERVFAIL) `tcpdump -i lo -n port 53 -vvv`:

```
00:54:24.526647 IP (tos 0x0, ttl 64, id 28229, offset 0, flags [none], proto UDP (17), length 80)
127.0.0.1.35017 > 127.0.0.1.53: [bad udp cksum 0xfe4f -> 0x3c85!] 42989+ [1au] A? example.com. ar: . OPT UDPsize=1232 [COOKIE c1e35a51547e4ea9] (52)
00:54:24.530333 IP (tos 0x0, ttl 64, id 38182, offset 0, flags [none], proto UDP (17), length 68)
127.0.0.1.53 > 127.0.0.1.35017: [bad udp cksum 0xfe43 -> 0x7bb5!] 42989 ServFail q: A? example.com. 0/0/1 ar: . OPT UDPsize=1232 (40)
```

(`dig @(external IP) example.com`, success) `tcpdump -i eth0 -n port 53 -vvv`:

```
01:02:15.420040 IP (tos 0x0, ttl 64, id 55959, offset 0, flags [none], proto UDP (17), length 80)
192.168.1.88.47723 > (external ip).53: [udp sum ok] 40442+ [1au] A? example.com. ar: . OPT UDPsize=1232 [COOKIE 9cdf698e58a68662] (52)
01:02:15.421020 IP (tos 0x0, ttl 64, id 42059, offset 0, flags [DF], proto UDP (17), length 84)
(external IP).53 > 192.168.1.88.47723: [udp sum ok] 40442 q: A? example.com. 1/0/1 example.com. [22m29s] A 93.184.215.14 ar: . OPT UDPsize=4096 (56)
```

**Comment by saloniamatteo** [comment]
I would also like to point out that this is a unique behavior I have never seen in my life -- I have other services listening on localhost, and they work fine!

P.S. Yes, I do have a firewall. No, it does not have weird rules that prevent localhost from talking to itself. Yes, I have checked.

**Comment by saloniamatteo** [comment]
Update: I've also tried to use [this config](https://github.com/MichaIng/DietPi/blob/master/.conf/dps_182/unbound.conf), however queries to localhost still contain SERVFAIL.

**Comment by saloniamatteo** [comment]
After multiple hours of debugging, I finally fixed the issue.

**TL;DR** Vodafone seems to be manipulating DNS queries.

***

For context, my ISP is Vodafone.

What happened is, since Vodafone has to "ensure" ""optimal"" service, they effectively try to cover up settings, saying they are managed by "Super Wi-Fi", requiring to download their app to manage router settings (except, there are no useful settings in that app...).

The settings are visible for a split-second when opening a relevant tab (e.g. DNS), then they are covered up by the "Super Wi-Fi" banner.

I set my own settings/bypassed this retarded behavior by logging onto my router's settings page, went to the DNS tab and, using inspect-element, removed `display: none` from various hidden elements, making them visible, effectively setting DNS servers (for all clients) from Vodafone's, to my own:

**DNS mode**: automatic --> **DNS mode**: manual
**DNS servers**:
- **Primary**: 1.1.1.1
- **Backup**: 1.0.0.1
- [...]

Here's how the settings look, after everything has been successfully changed: 
![Vodafone_DNS](https://github.com/user-attachments/assets/5beabc97-1b76-423e-9e37-aa0a21e78967)

***

I think Vodafone is manipulating DNS queries: all queries made to `example.com`, using ANY DNS Server (Cloudflare, Google, Quad9, etc.), were resulting in normal-looking responses... except the upstream DNS Servers were never actually queried, using Vodafone's DNS under the hood.

This is problematic for me, because I'm running an email server and, in order to query DNSBLs/URIBLs, most services require I run my own caching nameserver. What made me instantly realize the issue was this command:

`dig test.uribl.com.multi.uribl.com txt +short`

Why? Because it answers with my originating DNS IP:

```txt
"127.0.0.1 -> Query Refused. See http://uribl.com/refused.shtml for more information [Your DNS IP: x.x.x.x]"
```

Taking the "DNS IP" above led me to my ISP's nameservers, even if I DID NOT use them, and instead used Cloudflare's, Google's, and even the local unbound instance! (Values set both using `dig`'s `@` parameter, and set in `/etc/resolv.conf`)

Now, after changing the router config, here's how an healthy response looks like:

```
"permanent testpoint"
```

***

Apart from querying DNSBLs/URIBLs and whatnot, the most important issue is: Unbound does NOT answer queries from `localhost`.

Yes, you read correctly.. **localhost** out of all places! This further proved to me something sketchy was going on.

After changing the DNS setting in the router, querying `localhost` on port 53 works like a charm! No issues, finally!

**Comment by wcawijngaards** [comment]
I believe that Unbound was not answering queries from localhost, because the dig query was using port 53 (from the line `;; SERVER: ::1#53(::1) (UDP)`), and Unbound is configured to listen only for queries on port 8080. The `interface: lo@8080` causes it to not reply to dig with port 53 on localhost, another DNS service is responding, is asking the router. Possibly that is what caused the answers for the queries. Also port 8080 is not configured for HTTPS responses. Like `interface: lo@53` would cause unbound to attempt to listen to queries for localhost on port 53, but likely that will fail to setup the sockets due to another service already there.

**Comment by saloniamatteo** [comment]
@wcawijngaards, I'm sorry if I worded in a confusing way: the config file is an example, the same behavior was observed even if the interfaces were set to `lo` (with and without specifying port 53); I also believe I said the issue persisted with any internal IP address (127.0.0.1, ::1, 192.168.1.x, ...)

**Comment by wcawijngaards** [comment]
Perhaps your analysis is entirely correct, I was looking at the config statements and debug logs. The unbound logs could have said what results it got from querying its upstream, by enabling verbose logging, say verbosity level 4, it then prints a dig-style output for the upstream result and that could indicate what is going on, in terms of that failure causing servfail or it returning modified information from the ISP. But since it works already for you, I guess never mind. Also unbound has `log-servfail: yes` that prints short information on servfails, that could be useful for a future debug case :-) .
